src/blog/0.3/linear-types-borrowing.vmd

```diff
@@ -0,0 +1,673 @@
+
+
+But still, I wondered if it _could_ be done. Could we get rid of that last little bit of overhead?
+
+
+A few weeks later, I discovered the solution.
+
+
+<slice/>
+
+# Moving into and out of functions
+
+<<<<
+Instead of taking a `&Random` reference parameter like this:
+
+```
+func RandInt(rand &Random) i64 {
+  x = rand.num;
+  set x = x + (x / 200096i64);
+  set x = x - (x * 33554432i64);
+  set x = x + (x / 134217728i64);
+  set rand.num = x;
+  return x;
+}
+```
+////
+`RandInt` could take (and return) an owned `Random`, like this:
+
+```
+func RandInt(rand Random) (i64, Random) {
+  x = rand.num;
+  set x = x + (x / 200096i64);
+  set x = x - (x * 33554432i64);
+  set x = x + (x / 134217728i64);
+  set rand.num = x;
+  return (x, rand);
+}
+```
+>>>>
+
+
+Here's what we changed:
+
+ * `rand &Random` became `rand Random`. Instead of taking a reference, the caller moves the actual Random instance itself.
+ * `return x` became `return (x, rand);` which returns a two-element tuple, which also has the Random instance.
+ * The return type `i64` became `(i64, Random)` accordingly.
+
+
+Now, `rand.num` isn't dereferencing a reference, it's accessing a field of an owned value.
+
+Vale's compiler sees this, and doesn't insert a generation check, because it knows the object is alive.
+
+
+Of course, that requires we change how `MakeBoard` calls `RandInt`.
+
+
+<<<<
+It goes from this:
+
+```
+r = RandInt(rand);
+```
+
+where `rand` is a `&Random` reference...
+////
+...to this:
+
+```
+[r, set rand] = RandInt(rand);
+```
+
+where `rand` is an actual `Random`, not just a reference.
+>>>>
+
+
+Now, that line:
+
+ * Moves the `rand` instance into the `RandInt` call.
+ * Uses the square braces to break apart the returned tuple into its two elements:
+    * The `i64` part assigned to the local variable `r`.
+    * The `Random` part is put into the `rand` variable via `set rand`.
+
+
+Here's the rest of `MakeBoard`, for context.
+
+
+<<<<
+It went from this:
+
+```
+// Returns a 2D array of random bools.
+func MakeBoard(
+    rand &Random,
+    num_rows int,
+    num_cols int)
+[][]bool {
+  rows = [][]bool(num_rows);
+  foreach row_i in 0..num_rows {
+    row = []bool(num_cols);
+    foreach col_i in 0..num_cols {
+      r = RandInt(rand);
+      row.push((r mod 2) == 0);
+    }
+    rows.push(row);
+  }
+  return rows;
+}
+```
+////
+...to this:
+
+```
+// Returns a 2D array of random bools.
+func MakeBoard(
+   rand Random,
+   num_rows int,
+   num_cols int)
+([][]bool, Random) {
+  rows = [][]bool(num_rows);
+  foreach row_i in 0..num_rows {
+    row = []bool(num_cols);
+    foreach col_i in 0..num_cols {
+      [r, set rand] = RandInt(rand);
+      row.push((r mod 2) == 0);
+    }
+    rows.push(row);
+  }
+  return (rows, rand);
+}
+```
+>>>>
+
+
+Let's break that down:
+
+ * The `rand &Random` parameter became `rand Random`, to take an actual `Random` instance instead of a reference.
```

Comment:
"Again t" I didn't realize at first that we were talking about MakeBoard instead of reiterating how RandInt changed